src/main/scala/com/microsoft/hyperspace/index/rules/BaseRuleHelper.scala

```diff
@@ -0,0 +1,574 @@
+/*
+ * Copyright (2020) The Hyperspace Project Authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.microsoft.hyperspace.index.rules
+
+import scala.collection.mutable
+
+import org.apache.hadoop.fs.Path
+import org.apache.spark.sql.SparkSession
+import org.apache.spark.sql.catalyst.catalog.BucketSpec
+import org.apache.spark.sql.catalyst.expressions.{Attribute, AttributeReference, In, Literal, Not}
+import org.apache.spark.sql.catalyst.optimizer.OptimizeIn
+import org.apache.spark.sql.catalyst.plans.logical._
+import org.apache.spark.sql.execution.datasources._
+import org.apache.spark.sql.execution.datasources.parquet.ParquetFileFormat
+import org.apache.spark.sql.types.{LongType, StructType}
+
+import com.microsoft.hyperspace.Hyperspace
+import com.microsoft.hyperspace.index._
+import com.microsoft.hyperspace.index.IndexLogEntryTags.{HYBRIDSCAN_RELATED_CONFIGS, IS_HYBRIDSCAN_CANDIDATE}
+import com.microsoft.hyperspace.index.plans.logical.{BucketUnion, IndexHadoopFsRelation}
+import com.microsoft.hyperspace.index.sources.{FileBasedRelation, FileBasedSourceProviderManager}
+import com.microsoft.hyperspace.util.HyperspaceConf
+
+class BaseRuleHelper(spark: SparkSession) {
+
+  /**
+   * Filter the given candidate indexes by matching signatures and index status.
+   * If Hybrid Scan is enabled, it compares the file metadata directly, and does not
+   * match signatures. By doing that, we could perform file-level comparison between
+   * index source files and the input files of the given plan. If there are some common
+   * files, the index is considered as a candidate.
+   *
+   * @param indexes List of available indexes.
+   * @param relation Relation with which given indexes are compared.
+   * @return Active indexes built for this plan.
+   */
+  def getCandidateIndexes(
```

Comment:
Please address this comment - to reduce the diff

Reply:
@sezruby 

1) I rather not put back `getCandidateIndexes` as the method requires spark being given to it. The current implementation is consistent as the methods that don't require spark are in `RuleUtils` while the other are part of the helper instances.
2) There 3000 lined due to extensive tests that I added to the verify the plan transformation. In a way or another someone should review it.
3) "_Frankly speaking_", I already did a lot of extra work complying with your requests of splitting in multiple PRs. Now, once more this request to split PR to this same feature. This is very frustrating for me and looks like you using any opportunity to push this back. I would understand the split into multiple PRs for new features but I though we did come to an agreement at the time we decided to split the feature in 4-5 other PRs.

Reply:
1=>
It will be  refactored out after rule refactoring. Just put it back to reduce the diff...
You can use spark from FilterIndexRule / JoinIndexRule, as before.
Also no need to `getCandidate` function in this class. (IndexPlanApplyHelper ?) - they have different role.

2/3=>
I understand your frustration, but the changes modify wide range of codebase and also core part.
This refactoring was not considered at the time we agreed and that's why this change became huge.
And you know small & explicit changes can be merged quickly, in general.. and splitting PR is also helping to trace the history. 

Sorry but I think we should do that. If you are not up for it, I'll push a refactoring PR(no nested related code) based on your PR, until early next week.

Reply:
@sezruby I want to correct you. The code does not add lots of changes. If you remember - and to help you remember, here is a previous discussion a month back, https://github.com/microsoft/hyperspace/pull/380#issuecomment-820231714 - we agreed to have separate classes for the already existing code and the nested features. So now we have `BaseRuleHelper` and `NestedRuleHelper`. From the old `RuleUtils` I extracted all the methods that need `spark` into the `BaseRuleHelper` and kept the methods that don't need `spark` in `RuleUtils` as static methods. So, from the old `RuleUtils` now we have 2 files: a stripped down `RuleUtils` and the `BaseRuleHelper`. The classes have the code unchanged from the old `RuleUtils`, only the method signatures are changed now that `spark` is a property inside the instance.

The new things are inside `NestedRuleHelper` which extends some methods from `BaseRuleHelper` to accommodate the nested fields support.

This is exactly what we agreed on that comment.

There is no other refactor here than this what I explained above refactor that does NOT make any sense without the next changes.

I propose to merge try merge this and you create the changes that you want on top of it.

@imback82, @rapoth: What is your opinion?